File: frontend/lib/mock/playground.ts
import type { PlaygroundQueryInput, PlaygroundResponse } from "@/lib/types";
import { getMockDocument } from "@/lib/mock/documents";

export function buildMockPlaygroundResponse(
  input: PlaygroundQueryInput,
): PlaygroundResponse {
  const rewritten = `${input.query.trim()} (genişletildi: vektör araması, hibrit=${input.searchType})`;
  const doc = getMockDocument("doc-1");
  const chunks = [
    {
      id: "c-1",
      text: "Hibrit geri getirme, kurumsal geri çağırma için sıkı gömüler ile seyrek BM25’i birleştirir.",
      score: 0.91,
      rerankScore: input.searchType !== "bm25" ? 0.88 : undefined,
      documentName: doc?.name ?? "Q1-Architecture.pdf",
    },
    {
      id: "c-2",
      text: "İzlenebilirlik: alım adımlarını, sorgu izlerini ve hat başına gecikme dağılımını günlükler.",
      score: 0.84,
      rerankScore: input.searchType !== "bm25" ? 0.79 : undefined,
      documentName: doc?.name ?? "Q1-Architecture.pdf",
    },
    {
      id: "c-3",
      text: "VectorDock PDF/DOCX/TXT alır, yapılandırılabilir örtüşmeyle parçalar ve vektörleri depoya yazar.",
      score: 0.78,
      rerankScore: input.searchType !== "bm25" ? 0.74 : undefined,
      documentName: doc?.name ?? "Q1-Architecture.pdf",
    },
  ].slice(0, Math.min(input.topK, 3));

  const embedding = 42;
  const retrieval = input.searchType === "hybrid" ? 128 : 96;
  const reranking = input.searchType === "bm25" || !input.searchType ? 0 : 118;
  const llm = 512;
  const total = embedding + retrieval + reranking + llm;

  const finalPrompt = `Sistem: VectorDock asistanısın. Yalnızca BAĞLAM kullan.\nKullanıcı: ${input.query}\nBağlam:\n${chunks.map((c) => `- ${c.text}`).join("\n")}`;

  return {
    rawQuery: input.query,
    rewrittenQuery: rewritten,
    pipeline: [
      {
        id: "rewrite",
        name: "Sorgu yeniden yazma",
        status: "success",
        latencyMs: 18,
        input: input.query,
        output: rewritten,
      },
      {
        id: "embed",
        name: "Gömme",
        status: "success",
        latencyMs: embedding,
        input: rewritten,
        output: "[vektör boyut 3072]",
      },
      {
        id: "retrieve",
        name:
          input.searchType === "semantic"
            ? "Vektör araması"
            : input.searchType === "hybrid"
              ? "Hibrit (sıkı + BM25)"
              : "BM25",
        status: "success",
        latencyMs: retrieval,
        output: `${chunks.length} parça ≥ ${input.threshold}`,
      },
      {
        id: "rerank",
        name: "Yeniden sıralayıcı",
        status: input.searchType === "bm25" ? "skipped" : "success",
        latencyMs: reranking,
        output:
          input.searchType === "bm25"
            ? undefined
            : "çapraz kodlayıcı ile üst-k yeniden sıra",
      },
      {
        id: "prompt",
        name: "İstem oluşturma",
        status: "success",
        latencyMs: 12,
        output: `${finalPrompt.length} karakter`,
      },
      {
        id: "llm",
        name: "LLM",
        status: "success",
        latencyMs: llm,
        output: "tamamlama",
      },
    ],
    chunks,
    finalPrompt,
    answer: `Geri getirilen bağlama göre VectorDock, yapılandırılabilir top-k=${input.topK} ve eşik=${input.threshold} ile ${input.searchType} geri getirmeyi destekler. Özet: hibrit sıkı+seyrek geri çağırmayı iyileştirir; hat aşamaları gecikme izlenebilirliği için izlenir.`,
    tokens: { input: 420, output: 180, total: 600 },
    latencyBreakdown: {
      embedding,
      retrieval,
      reranking,
      llm,
      total,
    },
  };
}

File: app/schemas/playground.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

StepStatus = Literal["success", "skipped", "error"]


class PlaygroundQueryBody(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    query: str
    collection_id: str = Field(alias="collectionId")
    search_type: Literal["semantic", "hybrid", "bm25"] = Field(alias="searchType")
    top_k: int = Field(alias="topK")
    threshold: float
    ollama_model: str | None = Field(default=None, alias="ollamaModel")
    system_prompt: str | None = Field(default=None, alias="systemPrompt")
    user_prompt_with_context: str | None = Field(default=None, alias="userPromptWithContext")
    user_prompt_no_context: str | None = Field(default=None, alias="userPromptNoContext")


class PipelineStepOut(BaseModel):
    model_config = ConfigDict(ser_json_by_alias=True, populate_by_name=True)

    id: str
    name: str
    status: StepStatus
    latency_ms: int = Field(serialization_alias="latencyMs")
    input: str | None = None
    output: str | None = None


class RetrievedChunkOut(BaseModel):
    model_config = ConfigDict(ser_json_by_alias=True, populate_by_name=True)

    id: str
    text: str
    score: float
    rerank_score: float | None = Field(default=None, serialization_alias="rerankScore")
    document_name: str = Field(serialization_alias="documentName")


class LatencyBreakdownOut(BaseModel):
    model_config = ConfigDict(ser_json_by_alias=True, populate_by_name=True)

    embedding: int
    retrieval: int
    reranking: int
    llm: int
    total: int


class TokensOut(BaseModel):
    model_config = ConfigDict(ser_json_by_alias=True, populate_by_name=True)

    input: int
    output: int
    total: int


class PlaygroundResponseOut(BaseModel):
    model_config = ConfigDict(ser_json_by_alias=True, populate_by_name=True)

    raw_query: str = Field(serialization_alias="rawQuery")
    rewritten_query: str = Field(serialization_alias="rewrittenQuery")
    pipeline: list[PipelineStepOut]
    chunks: list[RetrievedChunkOut]
    final_prompt: str = Field(serialization_alias="finalPrompt")
    llm_model: str = Field(serialization_alias="llmModel")
    answer: str
    tokens: TokensOut
    latency_breakdown: LatencyBreakdownOut = Field(serialization_alias="latencyBreakdown")


def stub_playground_response(body: PlaygroundQueryBody) -> PlaygroundResponseOut:
    msg = "Playground sorgusu henüz uygulanmadı (vektör arama + LLM)."
    step = PipelineStepOut(
        id="stub",
        name="placeholder",
        status="skipped",
        latency_ms=0,
        output=msg,
    )
    return PlaygroundResponseOut(
        raw_query=body.query,
        rewritten_query=body.query,
        pipeline=[step],
        chunks=[],
        final_prompt="",
        llm_model="—",
        answer=msg,
        tokens=TokensOut(input=0, output=0, total=0),
        latency_breakdown=LatencyBreakdownOut(
            embedding=0,
            retrieval=0,
            reranking=0,
            llm=0,
            total=0,
        ),
    )
